src/lib/queries/lexical-query.js

```diff
@@ -52,21 +53,35 @@ export default class LexicalQuery extends Query {
   * iterations () {
     let formLexeme = new Lexeme(new Lemma(this.selector.normalizedText, this.selector.languageID), [])
     this.ui.updateLanguage(this.selector.languageID)
+    if (this.tbAdapter && this.selector.data.treebank && this.selector.data.treebank.word) {
+      this.annotatedHomonym = yield this.tbAdapter.getHomonym(this.selector.languageID, this.selector.data.treebank.word.ref)
+    }
     if (!this.canReset) {
       // if we can't reset, proceed with full lookup sequence
       this.homonym = yield this.maAdapter.getHomonym(this.selector.languageID, this.selector.normalizedText)
       if (this.homonym) {
+        if (this.annotatedHomonym) {
+          this.homonym.disambiguate(this.annotatedHomonym)
+        }
         this.ui.addMessage(this.ui.l10n.messages.TEXT_NOTICE_MORPHDATA_READY)
       } else {
-        this.ui.addImportantMessage(this.ui.l10n.messages.TEXT_NOTICE_MORPHDATA_NOTFOUND)
-        // Need to notify a UI controller that there is no morph data on this word in an analyzer
-        // However, controller may not have `morphologyDataNotFound()` implemented, so need to check first
-        if (this.ui.morphologyDataNotFound) { this.ui.morphologyDataNotFound(true) }
-        this.homonym = new Homonym([formLexeme], this.selector.normalizedText)
+        if (this.annotatedHomonym) {
+          this.homonym = this.annotatedHomonym
```

Comment:
I'm thinking maybe it's better for `homonym.disambiguate()` to return a disambiguated homonym? Then we can do assignments like `this.homonym = this.homonym.disambiguate(this.annotatedHomonym)`.

If the purpose of disambiguation is to create an homonym that is a product of some transmutation between homonym A and homonym B maybe it's even better to make a static function in Homonym that will allow to record it like: `this.homonym = Homonym.disambiguate(this.homonym, this.annotatedHomonym)`. It might be easier to understand, on my opinion, as we will see that a new homonym is a result of some transformation applied to the two existing ones (if I understand a business logic correctly)? With such format, we can also add more parameters to disambiguate between more homonyms if we need too (`Homonym.disambiguate(homonymA, homonymB, homonymC)`).

What do you think? (Just a question for discussion)

Reply:
@kirlat yes I think this is a very good suggestion. Thank you! I will make the static function.

Reply:
this change is now in https://github.com/alpheios-project/data-models/pull/40